### PR TITLE
[WIP] Don't filter cloud provider nodes. #45118

### LINKED ISSUE #45118: cloud providers get instance/vm calls filter on running status
If a node is attached to a cloud provider and the machine is in any state other than running/available/ready/whatever cloud provider uses to indicate ready, it will be removed from the kubernetes cluster.

Looking at the various cloud providers, some of them filter nodes based on the state whereas others do not.  The node controller has logic for handling a node that is not updating status and evicting pods/tainting the node as needed.

Is there a reason why the cloud providers only return instances/vms that are "ready" instead of returning any instance/vm that matches the requested name?  It seems like it might be more consistent to have the cloud providers only filter on name and allow the caller to determine what to do if the found instance isn't in a desired state.

@gmarek @justinsb

### PR DESCRIPTION
**Pull request by rrati**
fixes #45118

WIP.  Only fixed for aws and to be used as a point of discussion and help identify any potential issues with the change.

**Release note**:
If a cloud provider node moved to any state other than available/running then it would be removed from the list of nodes in kubernetes.  That is no longer the case.  The node will not be removed from the list of nodes in kubernetes now but will instead end up marked as any other node that can't be contacted.

### CODE REVIEW

**Comment by justinsb** [comment]
Still not clear what problem we are solving.

**Comment by gmarek** [comment]
@justinsb - I believe it's #45118.

**Comment by rrati** [comment]
@justinsb I thought the issue was discussed on #45188 ?  Not sure education is the only solution.

**Comment by justinsb** [comment]
I believe this PR does not handle a terminated instance correctly.

It also remains unclear to me what problem it is trying to solve.

**Comment by ncdc** [comment]
@justinsb what I wrote here: https://github.com/kubernetes/kubernetes/issues/45118#issuecomment-298928499

@smarterclayton could you please weigh in on this?

**Comment by rrati** [comment]
@justinsb What do you expect to happen when the node is terminated?  With this change, the node becomes NotReady when it is stopped, and removed from kube when it is terminated.

**Comment by rrati** [comment]
This change also makes aws act like gce.  In gce when a node is stopped it becomes NotReady but does not disappear from kube.  The node is only deleted from kube when the instance is terminated/deleted.

**Comment by smarterclayton** [comment]
I don't think stopping a node means (to any cloud administrator) "delete it because it's never coming back".

Use cases:

1. I need to audit a security break - DO NOT DELETE THE NODE
2. I want to restart a node and I leave it shut down for five minutes - DO NOT DELETE THE NODE

Deleting the node should be opt in, and even in that case it's very common that you will have some nodes you don't want to delete on shut down.

**Comment by ncdc** [comment]
cc'ing teams/people who have contributed and/or reviewed code to the cloud providers affected in this PR:

@kubernetes/sig-aws-misc @kubernetes/sig-openstack-pr-reviews @simon3z @jbeda @thockin @thommay @abithap @dagnello 

tl;dr For certain cloud providers, if a server is stopped (but not terminated/deleted) in the cloud provider, Kubernetes waits a very short amount of time, and then deletes the Node from etcd. For other cloud providers, it waits until the server is actually terminated/deleted before doing so. We are proposing that all cloud providers be consistent, and that they must wait for the server to be terminated/deleted before taking any action.

**Comment by justinsb** [comment]
This behaviour does not live up to the standards of our community.

**Comment by rrati** [comment]
@justinsb How so?

**Comment by ncdc** [comment]
@justinsb do you want to reword #45118 to be that issue, or close it and start a new one?

**Comment by justinsb** [comment]
@ncdc I'm inclined to say a new issue is going to be clearer for everyone, but I'm easy either way...

**Comment by rrati** [comment]
@justinsb IMO there's a larger issue to discuss and work through that isn't aws specific.  Right now different cloud providers treat a stopped instance differently.  Some delete the node from kube, whereas others keep it.  I think we would want a common response across all cloud providers and if so we need to decide what we want that response to be.

**Comment by ncdc** [comment]
New issue SGTM

**Comment by thockin** [comment]
https://en.wikipedia.org/wiki/Ship_of_Theseus

If turning off a node loses local disk state, I'd call it a new node. It's probably not OK for a pod to be restarted (assuming it tolerates an outage) on a node but its `emptyDir` volumes not survive.

**Comment by gnufied** [comment]
Also filtering for running nodes is causing us other bugs - https://github.com/kubernetes/kubernetes/pull/46463 

The problem with #46463 is - 

When a node is stopped for a bit and then `VerifyAttachedVolume` function fires - since it can't find the node anymore, it considers all volumes to the node as detached, which is not true. 

For #46463 - we are considering a localized approach but this appears to be a real problem.

**Comment by gnufied** [comment]
Problem is - for both AWS & GCE, a terminated node indeed detaches disks but stopped nodes don't. So if `VerifyAttachedVolume` fired in-between reboots, we have problem.

**Comment by wongma7** [comment]
Right, to add more context in case anyone cares, the main intent of my PR #46463 is to fix a specific attach/detach bug, so I only changed the behavior for `DisksAreAttached` and, not being in a position to decide what's at contention here, maintained the filter-by-running behavior elsewhere.

If it's eventually agreed that a stopped/rebooting AWS instance is still a node and a variation of this PR gets in, great, that also fixes my bug. But if we don't then I'd like for some variation of my PR to get in, because for the specific purpose of checking where disks are attached, stopped/rebooting instances *do* have to be considered. Basically I think this isn't true anymore, at least not in the eyes of the (cloudprovider-agnostic/unaware) attach/detach controller: "everything that involves a not-running state is a new k8s Node" @justinsb

**Comment by rrati** [comment]
From my testing on opensack, aws, and gce a stopped node does not lose local state.  I can stop an instance and start it back up and it will join the kubernetes pool just like a physical node.  Terminating a node definitely loses local state.

The exception to that is anything that might be storage on ephemeral storage.  AWS specifically warns you about that when stopping the instance, but I don't think kube uses ephemeral storage for anything by default.

**Comment by gmarek** [comment]
@justinsb - do you think it's possible to reach an agreement before 1.7 rolls out?

**Comment by ncdc** [comment]
@gmarek I'd say whatever we end up doing here is going to be big enough that we need adequate time to thoroughly discuss and form consensus. It would be nice to solve the underlying inconsistency for 1.8 if possible.

**Comment by gmarek** [comment]
Fair point.

**Comment by lpabon** [comment]
What steps do we need to take to come to a concensus? Should we discuss it at a SIG-meeting or over email, or on GitHub?

**Comment by gmarek** [comment]
I'd say that you should discuss this issue on @kubernetes/sig-aws-misc, but as I'm not an interested party you can safely ignore me:) If the decision would be that we want to change NodeController logic around Node deletion, then we'll need to have a larger meeting, probably in @kubernetes/sig-cluster-lifecycle-misc meeting.

**Comment by ncdc** [comment]
This is broader than just AWS. It potentially affects all cloud providers (ideally, they would converge on consistent behavior, so some would have to change).

**Comment by gmarek** [comment]
I'd guess cluster lifecycle then.

**Comment by gmarek** [comment]
@kubernetes/sig-cluster-lifecycle-feature-requests

**Comment by rrati** [comment]
/remove-lifecycle rotten

Keep this open until there is direction

**Comment by sjenning** [comment]
I would like to restart discussion on this.

PR https://github.com/kubernetes/kubernetes/pull/59931 was merged a few weeks ago to include stopped instances in Openstack in the query and prevent them from being deleted by the node controller.  I don't see an issue with doing it for AWS as well.

In our case, while the node does re-register, it loses any labelling that was done to the old Node resource.

I, for one, am in favor of including both stopped and running instances in the query.  In the case we have, the customer is using EBS-backed instances so the disk is persistent.

**Comment by sjenning** [comment]
Seems like those discussions are still in the design phase.  Do we think we can patch this up with the one-liner in AWS cloudprovider in the meantime and handle it at a higher level when that is agreed upon?

**Comment by derekwaynecarr** [comment]
Can we separate out the consistency cross cloud discussion from incremental improvement on just the AWS cloud provider specific behavior?  If a user stops and later starts a node, any transient node labels and/or taints applied to a node are lost.  The operator views this as data loss, and I am inclined to agree as the impact of incorrect labeling can have broader ramifications to the set of workloads scheduled to that node.

I feel like the code in this area is becoming very difficult to reason about.  For example, if we implemented the `InstanceShutdownByProviderID` operation in the AWS cloud provider, the existing node lifecycle controller would just taint the node and move on.  Since the operation is not implemented, it falls back to the existence check, which only returns true if the instance state is running.  As a user of the cloud provider libraries, this is feels really hard to reason about from a maintainers point of view.

From @justinsb original comment it sounds like he may be willing to make progress on AWS stop/restart behavior which feels like a good step forward for AWS specific users.  It's possible I am missing a key piece of information on why this would not be desired, so input appreciated.

/cc @justinsb @countspongebob @jsafrane

**Comment by rrati** [comment]
I have a concern moving forward without agreement on the end goal.  We could end up moving in the wrong direction and causing 2 disruptive behaviors.  If we just improve AWS so not filter nodes, that is a change to the user (likely positive imo).  If the end direction is decided to be something else then the user likely experience yet another jarring change.

Is there some agreement on a direction for how this problem should be handled?  Are multiple changes in this area acceptable?
